pkg/apis/batch/validation/validation.go

```diff
@@ -546,6 +546,43 @@ func ValidateJobTemplateSpec(spec *batch.JobTemplateSpec, fldPath *field.Path, o
 	return allErrs
 }
 
+func validateCompletions(spec, oldSpec batch.JobSpec, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+
+	isIndexedJob := spec.CompletionMode != nil && *spec.CompletionMode == batch.IndexedCompletion
+	atLeastOneCompletionsNil := spec.Completions == nil || oldSpec.Completions == nil
+	bothCompletionsNil := spec.Completions == nil && oldSpec.Completions == nil
+
+	// Non-indexed Jobs cannot have completions set at all.
+	if !isIndexedJob {
+		if !bothCompletionsNil {
+			return apivalidation.ValidateImmutableField(spec.Completions, oldSpec.Completions, fldPath)
+		}
+		// Non-indexed Job with both completions set as nil is valid
+		return allErrs
+	}
+
+	// Indexed Jobs cannot set completions to nil
+	if isIndexedJob && atLeastOneCompletionsNil {
+		// nil check already performed in validateJobSpec, no need
+		// to add another error
+		return allErrs
+	}
+
+	// Completions is mutable only for Indexed Jobs
+	mutatedCompletions := *spec.Completions != *oldSpec.Completions
+	if mutatedCompletions && !isIndexedJob {
+		return apivalidation.ValidateImmutableField(spec.Completions, oldSpec.Completions, fldPath)
```

Comment:
isIndexedJob is always true at this point, this is unreachable

Reply:
Makes sense, done.